src/workspace/discovery.rs

```diff
@@ -295,13 +305,56 @@ fn discover_workspace_config_files_for_single_dir(
         specifier.clone(),
         &opts.config_parse_options,
       )?);
-      let maybe_pkg_json = load_pkg_json_in_folder(&config_file.dir_path())?;
+
+      // see what config would be loaded if we just specified the parent directory
+      let natural_config_folder_result =
+        load_config_folder(file.parent().unwrap());
+      let matching_config_folder = match natural_config_folder_result {
+        Ok(Some(natual_config_folder)) => {
+          if natual_config_folder
+            .deno_json()
+            .is_some_and(|d| d.specifier == config_file.specifier)
+          {
+            Some(natual_config_folder)
+          } else {
+            None
+          }
+        }
+        Ok(None) | Err(_) => None,
+      };
+
+      let config_folder = match matching_config_folder {
+        Some(config_folder) => config_folder,
+        None => {
+          // when loading the directory we would have loaded something else, so
+          // don't try to load a workspace and don't store this information in
+          // the workspace cache
+          let config_folder =
+            ConfigFolder::Single(DenoOrPkgJson::Deno(config_file));
+          return Ok(ConfigFileDiscovery::Workspace {
+            maybe_vendor_dir: resolve_vendor_dir(
+              config_folder.deno_json().map(|d| d.as_ref()),
+              opts.maybe_vendor_override.as_ref(),
+            ),
+            workspace: new_rc(Workspace::new_single(config_folder)),
+          });
```

Comment:
This is a bug fix here. Basically, if someone specifies a path like `not-deno.json` then we shouldn't do workspace discovery because there's no way that this configuration file could be part of the workspace.

Additionally, if someone specifies a `deno.jsonc` file when there's a `deno.json` file, the `deno.json` file would be part of the workspace and not the `deno.jsonc`, so we should also not do workspace resolution in this case.

Reply:
Wow, good catch 👍